Refactor operator norm

jlpaca opened the pull request:
First part of a process towards restating `deriv.lean` in terms of bundled bounded linear maps. Supersedes #726 : @sgouezel has kindly pointed out that much of the work there has become outdated and/or duplicated; this is an attempt at cleaning up the mess I made.

Some names for theorems and definitions have been altered for adherence to conventions, or for brevity. Please see comments on #726 and commit messages in this PR for detailed accounts of changes.

TO CONTRIBUTORS:

Make sure you have:

  * [ ] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/naming.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/tests/tactics.lean)
  * [ ] make sure definitions and lemmas are put in the right files
  * [ ] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/code-review.md)

jlpaca committed:
style(analysis/normed_space/bounded_linear_map): minor edits

- extract some variables
- use consistent notation for hypotheses & use `let` instead of invoking equation compiler
- remove spurious imports

jlpaca committed:
refactor(analysis/normed_space/operator_norm): refactor operator norm

- rename `bounded_linear_maps` to `bounded_linear_map`.

- make some straightforward lemmas available to `simp`.

- change the definition of `operator_norm` to be the inf of all the bounds
  (prove that the image of the unit ball is bounded as a lemma).

- change the names of some theorems to be consistent with conventions elsewhere
  in mathlib: replace `bounded_by_*` with `le_*`, `operator_norm_homogeneous` with
  `operator_norm_smul`.

- remove some implicit arguments.

- additional miscellaneous lemmas: that the operator norm is submultiplicative, also
  lipschitz continuity.

jlpaca committed:
chore(analysis/normed_space/operator_norm): add attribution

jlpaca committed:
refactor(analysis/normed_space/operator_norm): rename `operator_norm` to `op_norm`

jlpaca committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into refactor-op-norm

sgouezel commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,174 @@
-
-def bounded_linear_maps : subspace k (E → F) :=
-{ carrier := {A : E → F | is_bounded_linear_map k A},
+def bounded_linear_map : subspace k (E → F) :=
```
The parameters k, E and F should be explicit in `bounded_linear_map` (for instance by writing them as `(k : Type*)` and `(E : Type*)...` on lines 19 and 20, and then adding a line `variables {k E F G}` after the definition). Then you can write `bounded_linear_map α β γ` on line 36.

sgouezel commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,174 @@
-local notation `L(` E `,` F `)` := @bounded_linear_maps k E F _ _ _
+namespace bounded_linear_map
+
+notation β ` →L[`:25 α `] ` γ := @bounded_linear_map α β γ _ _ _
```
Why don't you use `E`, `k` and `F` here?

sgouezel commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,174 @@
 
-@[simp] lemma bounded_linear_maps.zero_smul {A : L(E, F)} : (0 : k) • A = 0 :=
-by ext; simp
+@[simp] lemma coe_zero : (0 : E →L[k] F) = 0 := rfl
```
```suggestion
lemma coe_zero : ((0 : E →L[k] F) : E → F) = 0 := rfl
```
The current version does not do what you think: there is no coercion, it says that `0 = 0` in `E →L[k] F`. With the simp attribute, it breaks the proof of `zero_smul` below (for an obscure `erw` reason). Anyway, the only interesting property of the 0 function is that it maps anything to 0, and this is already taken care of by `zero_apply`, so I think the simp attribute on `coe_zero` is not useful and can be safely removed.

sgouezel commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,174 @@
 
-/-- Bounded linear maps are ... bounded -/
-lemma exists_bound (A : L(E,F)) : ∃ c, 0 < c ∧ ∀ x : E, ∥A x∥ ≤ c * ∥x∥ := A.property.bound
+@[simp] lemma zero_smul : (0 : k) • f = 0  := by ext; simp
```
The recommended style is now `by { ext, simp }` if there are no multiple goals. Here and several times below.

sgouezel commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,174 @@
+-- bounds is nonempty and bounded below.
+lemma bounds_nonempty {f : E →L[k] F} :
+  ∃ c, c ∈ { c | c ≥ 0 ∧ ∀ x, ∥f x∥ ≤ c * ∥x∥ } :=
+  let ⟨M, hMp, hMb⟩ := f.property.bound in ⟨M, le_of_lt hMp, hMb⟩
```
remove indent

sgouezel commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,174 @@
+
+lemma bounds_bdd_below {f : E →L[k] F} :
+  bdd_below { c | c ≥ 0 ∧ ∀ x, ∥f x∥ ≤ c * ∥x∥ } :=
+  ⟨0, λ _ ⟨hn, _⟩, hn⟩
```
remove indent

sgouezel reviewed (dismissed):
You always indent your proofs by at least two spaces with respect to the left margin. Normally, the left-most lines of a proof should not be indented.

jlpaca committed:
style(analysis/normed_space/operator_norm): style fixes

- use curly brackets for one-line tactic proofs without multiple goals.
- fix error in statement of `coe_zero`.
- adjust indentation.

jlpaca committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into refactor-op-norm

jlpaca committed:
fix(analysis/normed_space/operator_norm): fix namespace scope

- place several theorems under the `bounded_linear_map` namespace.
- remove spurious `noncomputable`.

jlpaca committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into refactor-op-norm

jlpaca committed:
Merge branch 'master' into refactor-op-norm

jlpaca committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into refactor-op-norm
